File: bindings/nodejs/examples/encode-to-mp4.ts
/**
 * Auto video encode -> fragmented MP4 — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `@mediaway/encoder` npm package exists yet. This
 * file shows the target ergonomics for a future Node.js binding over
 * Mediaway's C ABI (napi-rs native addon under the hood, wrapped in
 * idiomatic Node/TypeScript: typed configs, `Error` subclasses, explicit
 * `close()`). This is the Node.js host reaching native code through the C
 * ABI — not the browser/WASM host. See ../README.md and docs/spec/c-ffi.md.
 *
 * Mirrors examples/encode_to_mp4.rs: pick the best available OS/GPU H.264
 * encoder automatically (Zero-Copy GPU path preferred, CPU-upload fallback),
 * wire its output packets into a fragmented MP4 muxer, push 90 synthetic
 * grey NV12 frames (3 s @ 30 fps), and write the finished MP4 to disk.
 *
 * Run (once the real package exists):
 *   npx tsx encode-to-mp4.ts
 */

import { writeFile } from "node:fs/promises";
import {
  AutoVideoEncodeConfig,
  EncodeSession,
  EncoderUnavailableError,
  openAutoEncoder,
  type VideoFrame,
} from "@mediaway/encoder";

const WIDTH = 640;
const HEIGHT = 480;
const FPS = 30;
const SECONDS = 3;
const FRAME_COUNT = FPS * SECONDS;

/** Build one grey NV12 frame: width*height Y bytes, then width*height/2 interleaved UV bytes. */
function makeGreyNv12Frame(pts: number): VideoFrame {
  const ySize = WIDTH * HEIGHT;
  const uvSize = ySize / 2;
  const data = Buffer.alloc(ySize + uvSize, 0x80); // Y=128, U=128, V=128 everywhere

  return {
    pts,
    duration: 1,
    width: WIDTH,
    height: HEIGHT,
    pixelFormat: "nv12",
    data,
  };
}

async function main(): Promise<void> {
  // Defaults for H.264 at this resolution/framerate, then override bitrate.
  const config: AutoVideoEncodeConfig = {
    ...AutoVideoEncodeConfig.defaults("h264", WIDTH, HEIGHT, { num: 1, den: FPS }),
    bitrateBps: 2_000_000,
  };

  let encoder;
  try {
    // Tries the best available backend on this machine (GPU Zero-Copy first,
    // CPU-upload fallback) — throws if none exists here.
    encoder = await openAutoEncoder(config);
  } catch (err) {
    if (err instanceof EncoderUnavailableError) {
      console.error(`encode-to-mp4: no encoder backend available on this platform (${err.message})`);
      return;
    }
    throw err;
  }

  console.log("encode-to-mp4: running on this platform");

  const session = new EncodeSession(encoder);
  try {
    for (let pts = 0; pts < FRAME_COUNT; pts++) {
      await session.writeFrame(makeGreyNv12Frame(pts));
    }

    // Flushes the encoder, finalizes the muxer, and returns the complete MP4 bytes.
    const mp4Bytes = await session.finish();

    await writeFile("out.mp4", mp4Bytes);

    console.log(`encode-to-mp4: ${FRAME_COUNT} frames -> out.mp4 (${mp4Bytes.length} bytes)`);
  } finally {
    session.close();
  }
}

await main();

File: bindings/kotlin/examples/EncodeToMp4.kt
/*
 * Auto video encode -> fragmented MP4 — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `mediaway` Kotlin/Java package exists yet. This file
 * shows the target ergonomics for a future Kotlin binding over Mediaway's C ABI
 * (JNI under the hood, wrapped idiomatically: Closeable resources used with
 * `.use { }`, Kotlin exceptions instead of raw error codes). The API is also
 * meant to read naturally from plain Java, so no Kotlin-only surface (inline
 * value classes, etc.) is used in the public shape. See ../README.md and
 * docs/spec/c-ffi.md.
 *
 * Mirrors mediaway_encoder::auto: build a config, open the best available
 * H.264 encoder backend for this platform (Zero-Copy GPU path preferred, CPU
 * upload as fallback), push raw NV12 frames into an encode session, and get
 * back complete fragmented MP4 bytes — the session wires the encoder's output
 * packets into a container muxer internally, so the caller never sees packets
 * or muxer tracks directly.
 *
 * Run (once the real package exists):
 *     kotlinc EncodeToMp4.kt -include-runtime -d encode-to-mp4.jar
 *     java -jar encode-to-mp4.jar
 */

import io.mediaway.encoder.AutoVideoEncodeConfig
import io.mediaway.encoder.AutoVideoEncoder
import io.mediaway.encoder.EncodeSession
import io.mediaway.encoder.NoBackendException
import io.mediaway.encoder.VideoFrame
import java.io.File

private const val WIDTH = 640
private const val HEIGHT = 480
private const val FRAME_RATE_NUM = 30
private const val FRAME_RATE_DEN = 1
private const val BITRATE_BPS = 2_000_000L
private const val FRAME_COUNT = 90 // 3 s at 30 fps
private const val FRAME_DURATION = 1L // one tick at a 1/30 s time base

/** Build one solid-grey NV12 frame: Y=128 everywhere, U/V=128 everywhere. */
private fun greyNv12Frame(pts: Long): VideoFrame {
    val ySize = WIDTH * HEIGHT
    val uvSize = WIDTH * HEIGHT / 2
    val data = ByteArray(ySize + uvSize)
    data.fill(128.toByte())

    return VideoFrame(
        pts = pts,
        duration = FRAME_DURATION,
        width = WIDTH,
        height = HEIGHT,
        pixelFormat = VideoFrame.PixelFormat.NV12,
        data = data,
    )
}

fun main() {
    // -- 1. Describe what to encode: H.264 defaults for this resolution and
    //       frame rate, then override the bitrate. ------------------------
    val config = AutoVideoEncodeConfig
        .h264Defaults(width = WIDTH, height = HEIGHT, frameRateNum = FRAME_RATE_NUM, frameRateDen = FRAME_RATE_DEN)
        .withBitrateBps(BITRATE_BPS)

    // -- 2. Open the best available backend on this machine. This is
    //       inherently fallible: on a platform/GPU with no suitable H.264
    //       encoder yet, bail out gracefully instead of crashing. ---------
    val encoder = try {
        AutoVideoEncoder.open(config)
    } catch (e: NoBackendException) {
        println("encode_to_mp4: no H.264 encoder backend available on this platform: ${e.message}")
        return
    }

    // -- 3. Push frames through an encode session, then flush to fMP4 bytes --
    val mp4Bytes = encoder.use { enc ->
        EncodeSession(enc).use { session ->
            for (i in 0 until FRAME_COUNT) {
                session.writeFrame(greyNv12Frame(pts = i.toLong()))
            }
            session.finish()
        }
    }

    val outFile = File("out.mp4")
    outFile.writeBytes(mp4Bytes)

    println("encode_to_mp4: $FRAME_COUNT frames (${WIDTH}x$HEIGHT @ $FRAME_RATE_NUM fps) -> ${mp4Bytes.size} bytes of fMP4")
    println("encode_to_mp4: wrote ${outFile.absolutePath}")
}
